### PR TITLE
feat: datatrails: include metric prefix filter

### LINKED ISSUE #79597: Data Trails: Show categories on first page
Related to issue #79598 

There are 2 options for users to narrow down metrics by prefixes:

1. Filters (hoping that users use some sort of preset filters for labels to narrow down the number of prefix filters)

2. Faceted Search

Image

Figma file w/ both options: https://www.figma.com/file/k87knv2idoa0EIsLpQOWkR/Data-Trails?type=design&node-id=416%3A44231&mode=design&t=QNUOVenWOtsVDmMI-1

Okay with either but my 2c:

When we were looking into product analytics for the Prometheus metrics builder / explorer, we found it was roughly 50/50 on if a user would just look for a specific metric first or start filtering down by labels. I believe for some (not all?) users, it comes to not knowing what the label / values are part of the metric they are looking for.

So based on this, we could do some pre-set filters with the filter design.
- Testing strategy A: Stick with one design, get data on how often users filter with the pre-sets and see if we need to pivot
- Testing strategy B: Build both filters and faceted search and A/B test — either a thorough design prototype, or in code. We would A/B test (or build very thorough design prototype. Observation would be to see which version leads to more filtering (and if the pre-set filters are being used)

I’m okay with either but I think testing strategy B might be more conclusive although more costly in time.

See slack discussion: https://raintank-corp.slack.com/archives/C0630AYC4SY/p1702606210174199

### LINKED ISSUE #81594: Data Trails: Remove same metric under the related tab
See thread: https://raintank-corp.slack.com/archives/C0630AYC4SY/p1706294007299009

### PR DESCRIPTION
**Pull request by darrenjaneczek**
**What is this feature?**

- Addresses #79597 
- Fixes #81594 
  - https://github.com/grafana/grafana/pull/81316/commits/db8d5056d854f5d077683ef43d0dfacf5f423029

A category tree is derived from all metric names that fit the current metric search criteria, populating a prefix filter Cascader component.

image

This cascader components allows the user to browse through the remaining metrics through the name-based hierarchy, and select a prefix filter to further narrow down the displayed results.

If the search terms are changed to something too restrictive, an error message will appear:
image

The Cascader allows text input, which I was not able to disable. By default, it would report "No results" for anything typed in, so I enabled `allowCustomValue`. This gives the user the ability to type in their own custom prefix, even an invalid one. This was preferable to the confusing "No results" outcome.

image

image

**Which issue(s) does this PR fix?**:

- Resolves #79597 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by darrenjaneczek** [comment]
# TODO
- [ ] Have the state be stored and updated with trail history nodes when they are created and restored

**Comment by darrenjaneczek** [comment]
(working on conflicts. please stand by)

**Comment by darrenjaneczek** [comment]
> The cascader search doesn't work for me locally, it's supposed to filter the cascader options based on what we write in the input right?

Unfortunately not. The component doesn't seem to perform the search (or doesn't find results). 
So instead, I enabled the ability for the user to enter a custom value because I was unable to just disable typing..

I am planning to revisit the component itself to get a better idea of what's going on with it, and add a few features relevant to this.

**Comment by darrenjaneczek** [comment]
> The cascader search doesn't work for me locally, it's supposed to filter the cascader options based on what we write in the input right?

I found a way to fix the Cascader to work closer to the way we expect.
- https://github.com/grafana/grafana/pull/81316/commits/a91a0b3fcb5b67a99da431e798c2ee93ec746556

**Review by adrapereira — CHANGES_REQUESTED** [review]
This is a great change! It greatly improves the navigation of the available metrics.
Left a couple of comments around where to store the intermediate lists of metrics.

Also found that the UI hangs while searching with the latest PR, it would be nice if we could debounce the search call somehow but I won't block this PR with it.

File: public/app/features/trails/MetricSelectScene.tsx
Comment (darrenjaneczek) [reply public/app/features/trails/MetricSelectScene.tsx:48]: If they are in state, I can trust React will be using the current values when I call *.useState().

I had experienced subtle bugs previously of things not updating when accessing object properties.

I would have to verify if it works fine as class variables. Perhaps we can explore that after merge.

**Review by adrapereira — APPROVED** [review]

**Pull request merged**
